students/ArchWu/Session8/mailroom_oo.py

```diff
@@ -0,0 +1,192 @@
+    
+def main():
+    
+    prompt = dedent('''
+                Choose an action:
+
+                (1) - Send a Thank You
+                (2) - Create a Report
+                (3) - Send letters to everyone
+                (4) - Quit
+
+                > ''')
+    switch = {
+        "1": submenu,
+        "2": db.report,
+        "3": save_letters_to_disk,
+        "4": goodbye
+    }
+    
+    menu(switch, prompt)
+
+
+def menu(switch, prompt):
+    while True:
+        try:
+            choice = str(input(prompt)).strip().lower()          
+            if choice not in switch.keys():
+                print("choose again")
+            else:
+                if switch[choice]():
+                    break
+        except KeyError:
+            print("Error, choice invalid")
+        
+
+def submenu():
+    prompt = ("To send a thank you, select one:\n\n"
+                "(1) Update donor and send thank-you\n"
+                "(2) List all existing DONORS\n"
+                "(3) Return to main menu\n > ")
+    switch = {
+        "1": thank_you,
+        "2": db.list_donor,
+        "3": return_menu,
+    }
+    menu(switch, prompt)
+
+def goodbye():
+    sys.exit()
+
+def return_menu():
+    return True
+
+
+def thank_you():
+    """
+    Ask user for donation amount, and then add it  to the DB
+    """
+    # Now prompt the user for a donation amount to apply. Since this is
+    # also an exit point to the main menu, we want to make sure this is
+    # done before mutating the db.
+    print("in take_donation")
+    name = input("Enter a donor name (new or existing): \n >")
+    if name == 'quit': return
+    while True:
+        amount_str = input("Enter a donation amount (or <enter> to exit)> ").strip()
+        if not amount_str:
+            # if they provide no input, go back to previous menu
+            return
+        # Make sure amount is a valid amount before leaving the input loop
+        try:
+            amount = float(amount_str)
+            if math.isnan(amount) or math.isinf(amount) or round(amount, 2) == 0.00:
+                raise ValueError
+        except ValueError:
+            print("error: donation amount is invalid\n")
+            continue
+        else:
+            break
+
+    if normalize_name(name) not in db.donors.keys():
+        new_donor = Donor(name)
+        db.add_donor(new_donor)
+    db.donors[normalize_name(name)].donations.append(amount)
+
+    """Generate Letter Here"""
+    return
+
+def save_letters_to_disk():
+    temp_path = os.getcwd()
+
+    for donor in db.donors.items():
+        with open('{}/{}.txt'.format(temp_path, donor[1].name), 'w') as f:
+            num_donation = donor[1].total
+            greetings = 'Dear {},\n'.format(donor[1].name)
+            body = '\nThank you for your generous gift ${} to us!\n'.format(num_donation)
+            ending = '\nSincerely,\n  ABC foundations'
+            f.write(greetings + body + ending)
+    return
+
+def normalize_name(name):
```

Comment:
You already have this in yourDonor object, you can call it from there (Donor.normalize_name)